BR avrcp subunit info passthrough and Pass Through command reception on the AVRCP target

makeshi opened the pull request:
Implemented AVRCP TG functionality in the Bluetooth stack by  support for subunit Info and Pass Through command reception, along with shell integration for testing.

- Implements reception and response handling for Subunit Info commands.
- Adds a callback to handle Pass Through command reception, including operation ID, button state, and    optional payload.
-Introduces an API to send Pass Through responses from the TG side.
- Add shell integration for subunit Info and Pass Through command testing.

makeshi committed:
Bluetooth: AVRCP: Implement subunit message reception on the AVRCP TG

Add a new callback to support the situation when a subunit info command
is received, and add an API to respond to the subunit info command.

Signed-off-by: Make Shi <make.shi@nxp.com>

makeshi committed:
Bluetooth: Shell: AVRCP: support subunit info request and response

Add avrcp_subunit_info_req callback function in shell and a command to set
the subunit info response.

Signed-off-by: Make Shi <make.shi@nxp.com>

makeshi committed:
Bluetooth: AVRCP: Implement Pass Through command reception on TG

Add a new callback to support the situation when a Pass Through command
is received, and add an API to respond to the Pass Through command.

Signed-off-by: Make Shi <make.shi@nxp.com>

makeshi committed:
Bluetooth: shell: Add support for AVRCP TG passthrough

Add a callback function in shell and a new command for AVRCP TG
passthrough support

Signed-off-by: Make Shi <make.shi@nxp.com>

makeshi committed:
Bluetooth: AVRCP: Dropping the connection when error response failure

Disconnect the L2CAP connection when fails on sending an error response
(unit info, subunit info, or passthrough).

Signed-off-by: Make Shi <make.shi@nxp.com>

makeshi commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -634,7 +652,51 @@ static void avrcp_vendor_dependent_cmd_handler(struct bt_avrcp *avrcp, uint8_t t
+		LOG_ERR("Invalid subunit info command");
+		goto err_rsp;
+	}
+
```
page != AVRCP_SUBUNIT_PAGE page already have been checked

makeshi commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -634,7 +652,51 @@ static void avrcp_vendor_dependent_cmd_handler(struct bt_avrcp *avrcp, uint8_t t
+err_rsp:
+	err = bt_avrcp_send_subunit_info_err_rsp(avrcp);
+	if (err) {
+		LOG_ERR("Failed to send subunit info error response");
```
updated

makeshi commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -977,3 +1039,43 @@ int bt_avrcp_tg_send_unit_info_rsp(struct bt_avrcp_tg *tg, uint8_t tid,
+	param[1] = FIELD_PREP(GENMASK(7, 3), rsp->subunit_type |
+		   FIELD_PREP(GENMASK(2, 0), rsp->max_subunit_id));
+
+	net_buf_add_mem(buf, param, sizeof(param));
```
updated

makeshi commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -977,3 +1039,43 @@ int bt_avrcp_tg_send_unit_info_rsp(struct bt_avrcp_tg *tg, uint8_t tid,
+	param[1] = FIELD_PREP(GENMASK(7, 3), rsp->subunit_type |
+		   FIELD_PREP(GENMASK(2, 0), rsp->max_subunit_id));
+
+	net_buf_add_mem(buf, param, sizeof(param));
```
Created a common static internal function to enable code reuse.

lylezhu2012 commented on include/zephyr/bluetooth/classic/avrcp.h:
```diff
@@ -378,6 +408,37 @@ int bt_avrcp_tg_register_cb(const struct bt_avrcp_tg_cb *cb);
+ *
+ *  @return 0 in case of success or error code in case of error.
+ */
+int bt_avrcp_tg_send_passthrough_rsp(struct bt_avrcp_tg *tg, uint8_t tid, bt_avrcp_rsp_t result,
+				     bt_avrcp_opid_t opid, bt_avrcp_button_state_t state,
+				     const uint8_t *payload, uint8_t len);
```
```suggestion
int bt_avrcp_tg_send_passthrough_rsp(struct bt_avrcp_tg *tg, uint8_t tid, bt_avrcp_rsp_t result, struct net_buf *buf);
```
Add provide an API `struct net_buf *bt_avrcp_create_pdu(struct net_buf_pool *pool, k_timeout_t timeout);` for allocating net buffer.

And for passthrough_rsp, the format of the  `buf->data` is `struct bt_avrcp_passthrough_rsp`.

makeshi replied:
Have provided the api in https://github.com/zephyrproject-rtos/zephyr/pull/91939/commits/f4bb4f4a0991f792427bb04348ae19522972bf25 PR

lylezhu2012 commented on include/zephyr/bluetooth/classic/avrcp.h:
```diff
@@ -378,6 +408,37 @@ int bt_avrcp_tg_register_cb(const struct bt_avrcp_tg_cb *cb);
+ *
+ *  @return 0 in case of success or error code in case of error.
+ */
+int bt_avrcp_tg_send_subunit_info_rsp(struct bt_avrcp_tg *tg, uint8_t tid,
+				      struct bt_avrcp_subunit_info_rsp *rsp);
```
Ditto.

makeshi replied:
According to the AVRCP specification, if a device implements this profile, it shall always return a fixed value in the response frame: the subunit_type field must be set to PANEL, and the max_subunit_ID field must be set to 0. update function to int bt_avrcp_tg_send_subunit_info_rsp(struct bt_avrcp_tg *tg, uint8_t tid);

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -977,3 +1116,83 @@ int bt_avrcp_tg_send_unit_info_rsp(struct bt_avrcp_tg *tg, uint8_t tid,
 	return avrcp_send(tg->avrcp, buf);
 }
+
+int bt_avrcp_tg_send_subunit_info_rsp(struct bt_avrcp_tg *tg, uint8_t tid,
+				      struct bt_avrcp_subunit_info_rsp *rsp)
+{
+	struct net_buf *buf;
+	uint8_t param[BT_AVRCP_SUBUNIT_INFO_RSP_SIZE];
+
+	if ((tg == NULL) || (tg->avrcp == NULL) || (rsp == NULL)) {
+		return -EINVAL;
+	}
+
+	if (!IS_TG_ROLE_SUPPORTED()) {
+		return -ENOTSUP;
+	}
+
+	/* It should be always value 0 in the max_subunit_ID and PANEL in the subunit_type field */
+	if ((rsp->max_subunit_id != 0u) || (rsp->subunit_type != BT_AVRCP_SUBUNIT_TYPE_PANEL)) {
+		return -EINVAL;
+	}
+
+	buf = avrcp_create_subunit_pdu(tg->avrcp, tid, BT_AVCTP_RESPONSE, BT_AVRCP_RSP_STABLE);
+	if (!buf) {
+		LOG_WRN("Insufficient buffer");
+		return -ENOMEM;
+	}
+
+	avrcp_fill_subunit_info_param(param, rsp->subunit_type, rsp->max_subunit_id);
+
+	if (net_buf_tailroom(buf) < sizeof(param)) {
+		LOG_WRN("Not enough tailroom in buffer");
+		net_buf_unref(buf);
+		return -ENOMEM;
+	}
+
+	net_buf_add_mem(buf, param, sizeof(param));
+
+	return avrcp_send(tg->avrcp, buf);
+}
```
If there is a case that the response is not `STABLE`?

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -634,13 +670,116 @@ static void avrcp_vendor_dependent_cmd_handler(struct bt_avrcp *avrcp, uint8_t t
+		goto err_rsp;
+	}
+
+	if (buf->len < sizeof(*avrcp_hdr)) {
+		goto err_rsp;
+	}
```
I think we can add a filed `required_len` or `min_len` to `struct avrcp_handler`, that the received data length can be checked in common place instead of checking buffer length in each cmd/rsp `handler`.

makeshi committed:
Bluetooth: AVCTP: support fragmented packet transmission and reception

- Implemented fragmented packet handling for AVCTP protocol
- Refactored AVCTP header structure to support multiple packet types
- Enhanced AVCTP send logic with packet segmentation support
- Updated AVRCP integration to support new AVCTP fragmented flow

Signed-off-by: Make Shi <make.shi@nxp.com>

lylezhu2012 commented on subsys/bluetooth/host/classic/Kconfig:
```diff
@@ -466,6 +466,34 @@ config BT_AVCTP
+
+config BT_AVCTP_RX_DATA_BUF_SIZE
+	int "AVCTP RX Data Buffer Size"
+	default 1024
```
It is better to add a range of the size.

makeshi replied:
updated

lylezhu2012 commented on subsys/bluetooth/host/classic/Kconfig:
```diff
@@ -466,6 +466,34 @@ config BT_AVCTP
+
+config BT_AVCTP_TX_DATA_BUF_SIZE
+	int "AVCTP TX Data Buffer Size"
+	default 540
```
Why the rx size is 1024, while the tx size is 540?

makeshi replied:
I previously thought that the AVCTP transmission should use a 512-byte packet, because AVCRP assembles packets. and I realized that the TX and RX buffer sizes should be same. So, I plan to merge the two Kconfig options avctp_tx and avctp_rx to one Kconfig,

lylezhu2012 commented on subsys/bluetooth/host/classic/avctp.c:
```diff
@@ -75,51 +92,409 @@ static void avctp_l2cap_encrypt_changed(struct bt_l2cap_chan *chan, uint8_t stat
+		return 0;
+	}
+
+	SYS_SLIST_FOR_EACH_NODE(&avctp->rx_pending, node) {
```
```suggestion
	struct bt_avctp_frag_rx *rx, *next;
	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&avctp->rx_pending, rx, next, node) {
```

makeshi replied:
updated

lylezhu2012 commented on subsys/bluetooth/host/classic/avctp.c:
```diff
@@ -75,51 +92,409 @@ static void avctp_l2cap_encrypt_changed(struct bt_l2cap_chan *chan, uint8_t stat
+
+	pkt_type = BT_AVCTP_HDR_GET_PACKET_TYPE(&hdr->common);
+
+	if (pkt_type == BT_AVCTP_PKT_TYPE_SINGLE) {
+		hdr = net_buf_pull_mem(buf, BT_AVCTP_HDR_SIZE_SINGLE);
+		pid = hdr->pid;
+	} else {
+		return avctp_recv_fragmented(session, buf);
+
+	}
```
```suggestion
	if (pkt_type != BT_AVCTP_PKT_TYPE_SINGLE) {
		return avctp_recv_fragmented(session, buf);
	}
```

makeshi replied:
done

lylezhu2012 commented on subsys/bluetooth/host/classic/avctp.c:
```diff
@@ -30,11 +30,28 @@
+NET_BUF_POOL_FIXED_DEFINE(avctp_tx_pool, CONFIG_BT_AVCTP_TX_DATA_BUF_CNT,
+			  CONFIG_BT_AVCTP_TX_DATA_BUF_SIZE, CONFIG_BT_CONN_TX_USER_DATA_SIZE, NULL);
+
+static struct bt_avctp_tx avctp_tx[CONFIG_BT_AVCTP_TX_DATA_BUF_CNT*2];
+static K_FIFO_DEFINE(avctp_tx_free);
+
+static struct bt_avctp_frag_rx avctp_rx[CONFIG_BT_AVCTP_RX_DATA_BUF_CNT*2];
+static K_FIFO_DEFINE(avctp_rx_free);
```
There is a head room space `CONFIG_BT_AVCTP_TX_DATA_BUF_SIZE` for each buffer. So I think the avdtp can leverage this space. Than the `avctp_tx` and `avctp_rx` are not needed. It can make the fragment simply.

makeshi replied:
done, have used header room with reassembly_buf

makeshi commented on subsys/bluetooth/host/classic/Kconfig:
```diff
@@ -466,6 +466,34 @@ config BT_AVCTP
+
+config BT_AVCTP_TX_DATA_BUF_CNT
+	int "AVCTP TX Data Buffer Count"
+	default 3
```
updated

makeshi commented on subsys/bluetooth/host/classic/avctp.c:
```diff
@@ -75,51 +92,409 @@ static void avctp_l2cap_encrypt_changed(struct bt_l2cap_chan *chan, uint8_t stat
+		rx = bt_avctp_frag_rx_alloc();
+		if (rx == NULL) {
+			LOG_ERR("No rx buffers!");
+			net_buf_unref(buf);
```
Correct, buf owner should not here, fixed

makeshi commented on subsys/bluetooth/host/classic/avctp.c:
```diff
@@ -75,51 +92,409 @@ static void avctp_l2cap_encrypt_changed(struct bt_l2cap_chan *chan, uint8_t stat
+		return avctp_recv_fragmented(session, buf);
+
+	}
+	return dispatch_avctp_packet(session, buf, hdr, pid);
```
have applied if (pkt_type != BT_AVCTP_PKT_TYPE_SINGLE) suggestion, it will make code more clear

makeshi commented on subsys/bluetooth/host/classic/avctp.c:
```diff
@@ -75,51 +92,409 @@ static void avctp_l2cap_encrypt_changed(struct bt_l2cap_chan *chan, uint8_t stat
+	/* for SINGLE case*/
+	if (tx->pkt_type == BT_AVCTP_PKT_TYPE_SINGLE ||
+	   (tx->sent_len == 0 && tx->total_len <= max_payload_size)) {
+		err = send_fragmented_avctp(avctp, tx->tid, BT_AVCTP_PKT_TYPE_SINGLE,
```
Done,  thanks, have use a new send_single_avctp to send the original buf direclty

makeshi commented on subsys/bluetooth/host/classic/avctp.c:
```diff
@@ -75,51 +92,409 @@ static void avctp_l2cap_encrypt_changed(struct bt_l2cap_chan *chan, uint8_t stat
+
+			if (pkt_type == BT_AVCTP_PKT_TYPE_END) {
+				if (rx->number_packet > 0) {
+					LOG_WRN("Unexpected packets number: %d", rx->number_packet);
```
updated

makeshi commented on subsys/bluetooth/host/classic/avctp.c:
```diff
@@ -75,51 +92,409 @@ static void avctp_l2cap_encrypt_changed(struct bt_l2cap_chan *chan, uint8_t stat
+				return 0;
+			}
+		}
+	}
```
for  the matched tid and cr is not found in the avctp->rx_pending case, it will drop the received packet and add LOG_WRN it.  Two different fragmented packets of same avdtp connection can only be sent one by one, The interleaving of the fragments of different packets is not allowed.

MarkWangChinese replied:
Since two different fragmented packets of same avdtp connection can only be sent one by one, can this part of codes be optimized? If receiving one packet that is not the on-going fragments, I think the previous fragments should be discarded.

makeshi replied:
updated

makeshi commented on subsys/bluetooth/host/classic/shell/avrcp.c:
```diff
@@ -232,13 +238,49 @@ static void avrcp_set_browsed_player_req(struct bt_avrcp_tg *tg, uint8_t tid,
+	bt_avrcp_opid_t opid;
+	bt_avrcp_button_state_t state;
+
+	rsp = net_buf_pull_mem(buf, sizeof(*rsp));
```
have added a  new bt_avrcp_passthrough_cmd struct for it

makeshi commented on subsys/bluetooth/host/classic/shell/avrcp.c:
```diff
@@ -437,6 +479,132 @@ static int cmd_send_unit_info_rsp(const struct shell *sh, int32_t argc, char *ar
+		opid = BT_AVRCP_OPID_VENDOR_UNIQUE;
+	}
+
+	rsp->byte0 = FIELD_PREP(BIT(7), state) | FIELD_PREP(GENMASK(6, 0), opid);
```
Done , already provided a new Macro for it

lylezhu2012 reviewed (commented):
It is the TX Fragment comments.

makeshi commented on subsys/bluetooth/host/classic/avctp.c:
```diff
@@ -75,51 +97,380 @@ static void avctp_l2cap_encrypt_changed(struct bt_l2cap_chan *chan, uint8_t stat
+
+	user_data->sent_len += chunk_size;
+
+	if (user_data->sent_len == buf->len) {
```
updated now

makeshi commented on subsys/bluetooth/host/classic/avctp.c:
```diff
@@ -75,51 +97,380 @@ static void avctp_l2cap_encrypt_changed(struct bt_l2cap_chan *chan, uint8_t stat
+	if (avctp->reassembly_buf) {
+		net_buf_unref(avctp->reassembly_buf);
+	}
+	return -EINVAL;
```
OK, return 0 to keep L2CAP up

makeshi commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -61,6 +61,14 @@ struct avrcp_pdu_handler {
 #define AVRCP_BROW_AVCTP(_avctp) CONTAINER_OF(_avctp, struct bt_avrcp, browsing_session)
 
+NET_BUF_POOL_FIXED_DEFINE(avctp_ctrl_rx_pool, CONFIG_BT_MAX_CONN,
+			  BT_L2CAP_BUF_SIZE(BT_AVRCP_FRAGMENT_SIZE),
```
yes, right, updated
